plugins/tableselection/plugin.js

```diff
@@ -889,6 +858,76 @@
 		setTimeout( function() {
 			editor.fire( 'afterPaste' );
 		}, 0 );
+
+		function isCustomPaste( selection, selectedCells ) {
+			var ranges = selection.getRanges(),
+				table = ranges.length && ranges[ 0 ]._getTableElement( { table: 1 } ),
+				isIgnoredTable,
+				isBoundaryWithoutTable,
+				isCollapsedPartial,
+				isTableElement;
+
+			// Do not customize paste process in following cases:
+			// 1. No cells are selected.
+			if ( !selectedCells.length ) {
+				return false;
+			}
+
+			// 2. It's a boundary position but with no table pasted.
+			isBoundaryWithoutTable = ( boundarySelection && !pastedTable );
+
+			if ( isBoundaryWithoutTable ) {
+				return false;
+			}
+
+			// 2.1 If it's boundary position and table is pasted, proceed with custom paste.
+			if ( boundarySelection ) {
+				return true;
+			}
+
+			// 3. Content exceedes table (#875).
+			isTableElement = rangeContainsTableElement( ranges[ 0 ] );
+
+			if ( !isTableElement ) {
+				return false;
+			}
+
+			// 4. Table is ignoring tableselection (#2945).
+			isIgnoredTable = table && table.hasAttribute( ignoredTableAttribute );
+
+			if ( isIgnoredTable ) {
+				return false;
+			}
+
+			// 5. It's single range that does not fully contain table element and is not boundary, e.g. collapsed selection within
+			// cell, part of cell etc.
+			isCollapsedPartial = selectedCells.length === 1 &&
+				!isTableElement &&
```

Comment:
On one hand it is totally logical, but on the other hand [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351234420) we explicitly added unrequired parameters to function just to avoid misleading in the future. I'm afraid that if we remove conditions that are unnecessary with current check order here and [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351240007), we will lose the information that they are still important and we don't place them just to keep code DRY. We could keep information about omitted conditions in comment, but it won't be save neither as code may desynchronize with comments as you mentioned [here](https://github.com/ckeditor/ckeditor4/pull/3561#discussion_r351238455). WDYT?